hibernate-core/src/main/java/org/hibernate/engine/jdbc/internal/JdbcCoordinatorImpl.java

```diff
@@ -269,10 +269,7 @@ public void afterStatementExecution() {
 	@Override
 	public void afterTransaction() {
 		transactionTimeOutInstant = -1;
-		if ( getConnectionReleaseMode() == ConnectionReleaseMode.AFTER_STATEMENT ||
-				getConnectionReleaseMode() == ConnectionReleaseMode.AFTER_TRANSACTION ) {
-			this.logicalConnection.afterTransaction();
-		}
+		this.logicalConnection.afterTransaction();
```

Comment:
maybe I'm wrong but in this way I think we are releasing the connection after a transaction even when `ConnectionReleaseMode.ON_CLOSE`, probably we should just add `|| getConnectionReleaseMode() == ConnectionReleaseMode.BEFORE_TRANSACTION_COMPLETION`

Reply:
There's already a check in `this.logicalConnection.afterTransaction();`:

```
		if ( connectionHandlingMode.getReleaseMode() != ConnectionReleaseMode.ON_CLOSE ) {
			// NOTE : we check for !ON_CLOSE here (rather than AFTER_TRANSACTION) to also catch:
			// - AFTER_STATEMENT cases that were circumvented due to held resources
			// - BEFORE_TRANSACTION_COMPLETION cases that were circumvented because a rollback occurred
			//   (we don't get a beforeTransactionCompletion event on rollback).
			log.debug( "Initiating JDBC connection release from afterTransaction" );
			releaseConnection();
		}
```

Reply:
Ha but you're right, there's a call to super() which release resources (not the connection)... I'll change this.

Reply:
Fixed, thanks.